File: frontend-web/cartella-frontend/src/api/addressService.js
import api from './api';

const addressService = {
  // Get addresses by user ID
  getAddressesByUserId: async (userId) => {
    try {
      const response = await api.get(`/addresses/user/${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching addresses by user ID:', error);
      throw error;
    }
  },

  // Get addresses by user email
  getAddressesByEmail: async (email) => {
    try {
      const response = await api.get(`/addresses/user/email/${email}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching addresses by email:', error);
      throw error;
    }
  },

  // Get addresses by username
  getAddressesByUsername: async (username) => {
    try {
      const response = await api.get(`/addresses/user/username/${username}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching addresses by username:', error);
      throw error;
    }
  },

  // Create address for user by ID
  createAddress: async (userId, addressData) => {
    try {
      const response = await api.post(`/addresses/user/${userId}`, addressData);
      return response.data;
    } catch (error) {
      console.error('Error creating address:', error);
      throw error;
    }
  },

  // Create address for user by email
  createAddressByEmail: async (email, addressData) => {
    try {
      const response = await api.post(`/addresses/user/email/${email}`, addressData);
      return response.data;
    } catch (error) {
      console.error('Error creating address by email:', error);
      throw error;
    }
  },

  // Update address
  updateAddress: async (addressId, addressData) => {
    try {
      const response = await api.put(`/addresses/${addressId}`, addressData);
      return response.data;
    } catch (error) {
      console.error('Error updating address:', error);
      throw error;
    }
  },

  // Delete address
  deleteAddress: async (addressId) => {
    try {
      await api.delete(`/addresses/${addressId}`);
      return true;
    } catch (error) {
      console.error('Error deleting address:', error);
      throw error;
    }
  },

  // Set default address
  setDefaultAddress: async (addressId) => {
    try {
      const response = await api.put(`/addresses/${addressId}/default`);
      return response.data;
    } catch (error) {
      console.error('Error setting default address:', error);
      throw error;
    }
  }
};

export default addressService; 

File: backend/cartella/src/main/java/cit/edu/cartella/service/AddressService.java
package cit.edu.cartella.service;

import cit.edu.cartella.entity.Address;
import cit.edu.cartella.entity.User;
import cit.edu.cartella.repository.AddressRepository;
import cit.edu.cartella.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class AddressService {

    private final AddressRepository addressRepository;
    private final UserRepository userRepository;

    @Autowired
    public AddressService(AddressRepository addressRepository, UserRepository userRepository) {
        this.addressRepository = addressRepository;
        this.userRepository = userRepository;
    }

    public List<Address> getAddressesByUserId(Long userId) {
        return addressRepository.findByUserUserId(userId);
    }

    @Transactional
    public Address createAddress(Long userId, Address address) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            throw new RuntimeException("User not found");
        }

        User user = userOpt.get();
        address.setUser(user);
        
        // Set timestamps
        address.setCreatedAt(LocalDateTime.now());
        address.setUpdatedAt(LocalDateTime.now());
        
        // If this is the first address or isDefault is true, set it as default
        List<Address> existingAddresses = addressRepository.findByUserUserId(userId);
        if (existingAddresses.isEmpty() || address.isDefault()) {
            address.setDefault(true);
            
            // If this is set as default, unset any other default addresses
            if (!existingAddresses.isEmpty()) {
                for (Address existingAddress : existingAddresses) {
                    if (existingAddress.isDefault()) {
                        existingAddress.setDefault(false);
                        addressRepository.save(existingAddress);
                    }
                }
            }
        }
        
        return addressRepository.save(address);
    }

    @Transactional
    public Address updateAddress(Long addressId, Address addressDetails) {
        Optional<Address> addressOpt = addressRepository.findById(addressId);
        if (addressOpt.isEmpty()) {
            throw new RuntimeException("Address not found");
        }
        
        Address address = addressOpt.get();
        
        // Update fields
        if (addressDetails.getStreetAddress() != null) {
            address.setStreetAddress(addressDetails.getStreetAddress());
        }
        if (addressDetails.getCity() != null) {
            address.setCity(addressDetails.getCity());
        }
        if (addressDetails.getState() != null) {
            address.setState(addressDetails.getState());
        }
        if (addressDetails.getPostalCode() != null) {
            address.setPostalCode(addressDetails.getPostalCode());
        }
        if (addressDetails.getCountry() != null) {
            address.setCountry(addressDetails.getCountry());
        }
        
        // Update timestamp
        address.setUpdatedAt(LocalDateTime.now());
        
        return addressRepository.save(address);
    }

    @Transactional
    public Address setDefaultAddress(Long addressId) {
        Address address = addressRepository.findById(addressId)
            .orElseThrow(() -> new RuntimeException("Address not found"));

        Long userId = address.getUser().getUserId();
        List<Address> userAddresses = addressRepository.findByUserUserId(userId);

        // Set all addresses to non-default first
        for (Address addr : userAddresses) {
            addr.setDefault(false);
        }

        // Set the selected address as default
        address.setDefault(true);
        address.setUpdatedAt(LocalDateTime.now());

        // Save all changes in a single transaction
        addressRepository.saveAll(userAddresses);
        
        // Log for debugging
        System.out.println("Set default address: " + addressId);
        System.out.println("Total addresses updated: " + userAddresses.size());
        
        // Return the updated default address
        return addressRepository.findById(addressId)
            .orElseThrow(() -> new RuntimeException("Address not found after update"));
    }

    public void deleteAddress(Long addressId) {
        addressRepository.deleteById(addressId);
    }
} 